src/background.ts

```diff
@@ -0,0 +1,139 @@
+  preload: boolean;
+};
+
+type Fetching = {
+  fetching: true;
+};
+
+type SecurityHeaders = {
+  fetching: false;
+  hsts?: HSTS;
+};
+
+type SecurityHeadersOrFetching = SecurityHeaders | Fetching;
+
+const parseHSTS = (hsts: string): HSTS | undefined => {
+  const directives = hsts.replace(/[ "]/g, '').toLowerCase().split(';');
+
+  const maxAge = directives.find((d) => d.startsWith('max-age='))?.slice(8);
+  if (maxAge === undefined) return;
+
+  return {
+    maxAge: Number.parseInt(maxAge),
+    includeSubDomains: directives.includes('includesubdomains'),
+    preload: directives.includes('preload'),
+  };
+};
+
+const fetchSecurityHeaders = async (
+  url: string,
+  cacheOnly: boolean,
+): Promise<SecurityHeadersOrFetching | undefined> => {
+  const { host } = new URL(url);
+
+  const cache = await getCache(host);
+  if (cache !== undefined || cacheOnly) return cache;
+
+  await setCache(host, { fetching: true });
+
+  let headers: Headers;
+  try {
+    const res = await fetch(url, {
+      method: 'HEAD',
+      cache: 'no-store',
+      redirect: 'manual',
+      credentials: 'omit',
+    });
+    headers = res.headers;
+  } catch (e) {
+    console.warn(e);
+    await removeCache(host);
+    return;
+  }
+
+  const hsts = headers.get('strict-transport-security');
+
+  return await setCache(host, {
+    fetching: false,
+    hsts: hsts ? parseHSTS(hsts) : undefined,
+  });
+};
+
+const getCache = async (
+  host: string,
+): Promise<SecurityHeadersOrFetching | undefined> => {
+  return (await browser.storage.session.get(host))[host];
+};
+
+const setCache = async <T extends SecurityHeadersOrFetching>(
+  host: string,
+  cache: T,
+): Promise<T> => {
+  await browser.storage.session.set({ [host]: cache });
+  return cache;
+};
+
+const removeCache = async (host: string) => {
+  await browser.storage.session.remove(host);
+};
+
+const setBadge = async (tabId: number, secHeaders: SecurityHeaders) => {
+  const secure = secHeaders.hsts !== undefined && secHeaders.hsts.maxAge > 0;
+
+  await browser.action.setBadgeText({
+    text: secure ? '✓' : '✕',
+    tabId,
+  });
+  browser.action.setBadgeTextColor({
+    color: secure ? '#28a745' : '#dc3545',
+    tabId,
+  });
+};
+
+browser.webRequest.onResponseStarted.addListener(
+  async ({ url, method, responseHeaders, fromCache, tabId }) => {
+    if (responseHeaders === undefined) return;
+
+    if (fromCache) {
+      const secHeaders = await fetchSecurityHeaders(url, method !== 'GET');
+      if (secHeaders === undefined || secHeaders.fetching) return;
+      await setBadge(tabId, secHeaders);
+      return;
+    }
+
+    const header = responseHeaders.find(
+      ({ name }) => name.toLowerCase() === 'strict-transport-security',
+    )?.value;
+
+    await setCache(new URL(url).host, {
+      fetching: false,
+      hsts: header ? parseHSTS(header) : undefined,
+    });
+  },
+  {
+    urls: ['https://*/*'],
+    types: ['main_frame'],
+  },
+  ['responseHeaders'],
+);
```

Comment:
I use `chrome.storage.session.set()` in `setCache` and `chrome.action.setBadgeText()` in `setBadge`, are there any possibility of exceptions in these functions?

Reply:
I added handling of exceeding quota with commit b9ea886267b4541dec8ea0760a06b3086b4b1d6a. Errors during fetch are already handled in the `fetchSecurityHeaders` function. `chrome.action.setBadgeText()` gives errors with permissions and references to non-existent tabs, but I leave it as is because I have no options to deal with it.

Reply:
When an error occurs, it is automatically output to the console log, but is there any advantage to explicitly use `console.error()` etc.?

Reply:
Thank you. I think error logging is appropriate on the outer layer of the stack, so I will consider implementing uniform handling as I add more features.